include/exadg/operators/operator_base.cpp

```diff
@@ -627,6 +585,40 @@ OperatorBase<dim, Number, n_components>::initialize_block_diagonal_preconditione
     iterative_solver_data);
 }
 
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::update_block_diagonal_preconditioner_matrix_free() const
+{
+  elementwise_preconditioner->update();
+}
+
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::initialize_block_diagonal_preconditioner_matrix_based()
+  const
+{
+  // allocate memory
+  auto dofs =
+    matrix_free->get_shape_info(this->data.dof_index).dofs_per_component_on_cell * n_components;
+  matrices.resize(matrix_free->n_cell_batches() * vectorization_length, LAPACKMatrix(dofs, dofs));
+
+  // compute and factorize matrices
+  update_block_diagonal_preconditioner_matrix_based();
+}
+
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::update_block_diagonal_preconditioner_matrix_based() const
+{
+  // clear matrices
+  initialize_block_jacobi_matrices_with_zero(matrices);
+
+  // compute block matrices and add
+  add_block_diagonal_matrices(matrices);
+
+  calculate_lu_factorization_block_jacobi(matrices);
```

Comment:
It has been like this before the PR, but in general we should aim to combine the initialization, computation/add, and factorization of a single matrix at a time, rather than doing it globally. In accordance with my other comment, this hints at what kind of data structures we might want to prefer in a future design.